GitHubGraphQl.ps1

```diff
@@ -0,0 +1,363 @@
+                    PowerShell 6+ returns a 'System.Net.Http.HttpResponseMessage' exception response object.
+                #>
+
+                $requestId = ''
+
+                if ($ex.Response.PSTypeNames[0] -eq 'System.Net.Http.HttpResponseMessage')
+                {
+                    $requestId = ($ex.Response.Headers | Where-Object -Property Key -eq 'X-GitHub-Request-Id').Value
+                }
+                elseif ($ex.Response.PSTypeNames[0] -eq 'System.Net.HttpWebResponse')
+                {
+                    if ($ex.Response.Headers.Count -gt 0 -and
+                        -not [System.String]::IsNullOrEmpty($ex.Response.Headers['X-GitHub-Request-Id']))
+                    {
+                        $requestId = $ex.Response.Headers['X-GitHub-Request-Id']
+                    }
+                }
+
+                if (-not [System.String]::IsNullOrEmpty($requestId))
+                {
+                    Write-Debug -Message "GitHub RequestID '$requestId' in response header"
+
+                    $localTelemetryProperties['RequestId'] = $requestId
+                    $requestIdMessage += "RequestId: $requestId"
+                    $errorMessage += $requestIdMessage
+
+                    Write-Log -Message $requestIdMessage -Level Verbose
+                }
+            }
+
+            $newErrorRecordParms = @{
+                ErrorMessage = $errorMessage -join [Environment]::NewLine
+                ErrorId = $_.FullyQualifiedErrorId
+                ErrorCategory = $_.CategoryInfo.Category
+                TargetObject = $Body
+            }
+            $errorRecord = New-ErrorRecord @newErrorRecordParms
+
+            Write-Log -Exception $errorRecord -Level Error
+            Set-TelemetryException -Exception $ex -ErrorBucket $errorBucket -Properties $localTelemetryProperties
+
+            $PSCmdlet.ThrowTerminatingError($errorRecord)
+        }
+        else
+        {
+            Write-Debug -Message "Processing Other Exception '$($ex.PSTypeNames[0])'"
+
+            $newErrorRecordParms = @{
+                ErrorMessage = $ex.Message
+                ErrorId = $_.FullyQualifiedErrorId
+                ErrorCategory = $_.CategoryInfo.Category
+                TargetObject = $body
+            }
+            $errorRecord = New-ErrorRecord @newErrorRecordParms
+
+            Write-Log -Exception $errorRecord -Level Error
+            Set-TelemetryException -Exception $ex -ErrorBucket $errorBucket -Properties $localTelemetryProperties
+
+            $PSCmdlet.ThrowTerminatingError($errorRecord)
+        }
+    }
+    finally {
+        Write-Debug -Message "Processing Invoke-WebRequest 'finally' block"
+
+        # Restore original security protocol
+        [Net.ServicePointManager]::SecurityProtocol = $originalSecurityProtocol
+
+        # Record the telemetry for this event.
+        $stopwatch.Stop()
```

Comment:
Not sure I follow. If I put it in the `try` block and there is an exception before it, then this code will never get executed. Please give me detail on what you want me to do.

Reply:
Exactly.  If you look at the corresponding code in [`Invoke-GHRestMethod`](https://github.com/microsoft/PowerShellForGitHub/blob/a2329d67302c55c42bae7fab2c78e63e1abdb656/GitHubCore.ps1#L320-L326), the stopping of the timer and the setting of the telemetry event happen within the `try`.  From the way we've implemented our telemetry, the setting of the telemetry event represents a successful event, and the setting of the telemetry exception represents a failure for some reason.  We want to prevent the scenario where a call to `Invoke-GHRestMethod` or `Invoke-GHGraphQl` results in _both_ a telemetry event _and_ an exception for the majority case.

So, the specific request here is to move 

```powershell
        # Record the telemetry for this event.
        $stopwatch.Stop()
 
        if (-not [String]::IsNullOrEmpty($TelemetryEventName))
        {
            $telemetryMetrics = @{ 'Duration' = $stopwatch.Elapsed.TotalSeconds }
            Set-TelemetryEvent -EventName $TelemetryEventName -Properties $localTelemetryProperties -Metrics $telemetryMetrics
        }
```

from the `finally` to the `try` after the `Invoke-WebRequest`, similar to how it is in `Invoke-GHRestMethod`.